server/src/imfs/imfs.rs

```diff
@@ -0,0 +1,280 @@
+    ///
+    /// This idea corresponds to whether a file change event should result in
+    /// tangible changes to the in-memory filesystem. If a path would be
+    /// resident, we need to read it, and if its contents were known before, we
+    /// need to update them.
+    fn would_be_resident(&self, path: &Path) -> bool {
+        if self.inner.contains_key(path) {
+            return true;
+        }
+
+        if let Some(parent) = path.parent() {
+            if let Some(ImfsItem::Directory(dir)) = self.inner.get(parent) {
+                return !dir.children_enumerated;
+            }
+        }
+
+        false
+    }
+
+    /// Attempts to read the path into the `Imfs` if it doesn't exist.
+    ///
+    /// This does not necessitate that file contents or directory children will
+    /// be read. Depending on the `ImfsFetcher` implementation that the `Imfs`
+    /// is using, this call may read exactly only the given path and no more.
+    fn read_if_not_exists(&mut self, path: &Path) -> FsResult<()> {
+        if !self.inner.contains_key(path) {
+            let item = self.fetcher.read_item(path)
+                .map_err(|err| FsError::new(err, path.to_path_buf()))?;
+            self.inner.insert(path.to_path_buf(), item);
+        }
+
+        Ok(())
+    }
+}
+
+/// A reference to file or folder in an `Imfs`. Can only be produced by the
+/// entry existing in the Imfs, but can later point to nothing if something
+/// would invalidate that path.
+///
+/// This struct does not borrow from the Imfs since every operation has the
+/// possibility to mutate the underlying data structure and move memory around.
+pub struct ImfsEntry {
+    path: PathBuf,
+    is_file: bool,
+}
+
+impl ImfsEntry {
+    pub fn path(&self) -> &Path {
+        &self.path
+    }
+
+    pub fn contents<'imfs>(
+        &self,
+        imfs: &'imfs mut Imfs<impl ImfsFetcher>,
+    ) -> FsResult<&'imfs [u8]> {
+        imfs.get_contents(&self.path)
+    }
+
+    pub fn children(
+        &self,
+        imfs: &mut Imfs<impl ImfsFetcher>,
+    ) -> FsResult<Vec<ImfsEntry>> {
+        imfs.get_children(&self.path)
+    }
+
+    pub fn is_file(&self) -> bool {
+        self.is_file
+    }
+
+    pub fn is_directory(&self) -> bool {
+        !self.is_file
+    }
+}
+
+/// Internal structure describing potentially partially-resident files and
+/// folders in the `Imfs`.
+pub enum ImfsItem {
+    File(ImfsFile),
+    Directory(ImfsDirectory),
+}
+
+impl ImfsItem {
+    fn path(&self) -> &Path {
+        match self {
+            ImfsItem::File(file) => &file.path,
+            ImfsItem::Directory(dir) => &dir.path,
+        }
+    }
+}
+
+pub struct ImfsFile {
+    pub(super) path: PathBuf,
+    pub(super) contents: Option<Vec<u8>>,
+}
+
+pub struct ImfsDirectory {
+    pub(super) path: PathBuf,
+    pub(super) children_enumerated: bool,
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    use maplit::hashmap;
+
+    use super::super::noop_fetcher::NoopFetcher;
```

Comment:
> `use super::super`

Is this something common? O_o

Reply:
It's pretty awkward. It's referring to the `imfs` parent module, which is more difficult because all the tests are in a module module even deeper than the code it's testing...